# Lesson 0007 - breast cancer support vector machines classification
We assume, that the reader has worked through all the previous lessons.<br>
In this lesson, we perform on the breast cancer data set from [lesson 0006](https://github.com/Mathhead/Lessons-in-Machine-Learning/blob/master/lesson_0006_breast_cancer_linear_classifier_fine_tuning.ipynb) classification using support vector machines.<br>
The preparation of the data can be stolen from lesson 0006.<br>
Please note, that the support vector machine expects 1-dimensional classification data, therefore __train_y__ and __test_y__ need to be 1-dimensional.

In [1]:
import sklearn as sklearn
from sklearn.datasets import load_breast_cancer as lbc

print( sklearn.__version__ )

0.19.1


In [2]:
data = lbc()

data_x = data.data

data_y = data.target





import numpy as np

np.random.seed( 1234567890 )

print( np.__version__ )

1.14.3


In [3]:
train_x = np.zeros( shape = [ 400, 30 ] )

train_y = np.zeros( shape = [ 400 ] )

test_x = np.zeros( shape = [ 169, 30 ] )

test_y = np.zeros( shape = [ 169 ] )



random_integers = np.random.choice( range( 569 ), 400, replace = False )

random_integers = np.sort( random_integers )



j = 0

k = 0



for i in range( 569 ):
    
    if i == random_integers[ j ]:
        
        train_x[ j, : ] = data_x[ i, : ]
        
        train_y[ j ] = data_y[ i ]
        
        j = j + 1
        
        if j == 400:
            
            j = 0
            
    else:
        
        test_x[ k, : ] = data_x[ i, : ]
        
        test_y[ k ] = data_y[ i ]
        
        k = k + 1
        
        
        
        
mu = np.zeros( shape = [ 30 ] )

sigma = np.zeros( shape = [ 30 ] )




for i in range( 30 ):
    
    mu[ i ] = np.mean( train_x[ :, i ] )
    
    sigma[ i ] = np.std( train_x[ :, i ] )
    
    
    
    
for i in range( 30 ):
    
    train_x[ :, i ] = ( train_x[ :, i ] - mu[ i ] ) / sigma[ i ]
    
    test_x[ :, i ] = ( test_x[ :, i ] - mu[ i ] ) / sigma[ i ]

Now we steal the code for the support vector machine from [lesson 0002](https://github.com/Mathhead/Lessons-in-Machine-Learning/blob/master/lesson_0002_support_vector_machines_classification.ipynb).

In [4]:
from sklearn.svm import SVC as SVC

We use __SVC__ straight out of the box.<br>
We train the model __svc__, store the correctly predicted data in __hit__ and compute the accuracy of that model in __accuracy__.

In [5]:
svc = SVC()

svc.fit( train_x, train_y )

hit = ( svc.predict( test_x ) == test_y )

accuracy = 0.0

for i in range( 50 ):
    
    if hit[ i ]:
        
        accuracy = accuracy + 1.0
        
accuracy = accuracy * 2.0

print( 'The out of the box model reached an accuracy of ' + str( accuracy ) + '%' )

The out of the box model reached an accuracy of 96.0%


With an accuarcy of $96\%$ and a test set size of less than $170$, we have less than $7$ items missclassified, so there is hardly room for improvement. Therefore<br>
Class dismissed.